src/OrbitGl/TimeGraph.cpp

```diff
@@ -437,64 +438,163 @@ void TimeGraph::SelectAndMakeVisible(const TimerInfo* timer_info) {
+  if (type == orbit_client_data::ScopeType::kInvalid) return nullptr;
+
+  if (type == orbit_client_data::ScopeType::kDynamicallyInstrumentedFunction ||
+      type == orbit_client_data::ScopeType::kApiScope) {
+    return FindPreviousThreadTrackTimer(scope_id, current_time, thread_id);
+  }
+
   const TimerInfo* previous_timer = nullptr;
   uint64_t goal_time = std::numeric_limits<uint64_t>::lowest();
 
-  std::vector<const TimerInfo*> timers =
-      capture_data_->GetAllScopeTimers(std::numeric_limits<uint64_t>::lowest(), current_time);
+  std::vector<const TimerInfo*> timers = capture_data_->GetAllScopeTimers(
+      {type}, std::numeric_limits<uint64_t>::lowest(), current_time);
   for (const TimerInfo* current_timer : timers) {
-    if ((thread_id && thread_id.value() != current_timer->thread_id()) ||
-        capture_data_->ProvideScopeId(*current_timer) != scope_id) {
-      continue;
-    }
-    if ((current_timer->end() < current_time) && (goal_time < current_timer->end())) {
-      previous_timer = current_timer;
-      goal_time = current_timer->end();
+    if (ThreadMatches(thread_id, current_timer) &&
+        capture_data_->ProvideScopeId(*current_timer) == scope_id) {
+      UpdatePreviousTimerAndGoalTime(&previous_timer, goal_time, current_timer, current_time);
     }
   }
   return previous_timer;
 }
 
 const TimerInfo* TimeGraph::FindNextScopeTimer(uint64_t scope_id, uint64_t current_time,
                                                std::optional<uint32_t> thread_id) const {
+  const orbit_client_data::ScopeType type = capture_data_->GetScopeInfo(scope_id).GetType();
+  if (type == orbit_client_data::ScopeType::kInvalid) return nullptr;
+
+  if (type == orbit_client_data::ScopeType::kDynamicallyInstrumentedFunction ||
+      type == orbit_client_data::ScopeType::kApiScope) {
+    return FindNextThreadTrackTimer(scope_id, current_time, thread_id);
+  }
+
   const TimerInfo* next_timer = nullptr;
   uint64_t goal_time = std::numeric_limits<uint64_t>::max();
-  std::vector<const TimerInfo*> timers = capture_data_->GetAllScopeTimers(current_time);
+
+  std::vector<const TimerInfo*> timers = capture_data_->GetAllScopeTimers({type}, current_time);
   for (const TimerInfo* current_timer : timers) {
-    if ((thread_id && thread_id.value() != current_timer->thread_id()) ||
-        capture_data_->ProvideScopeId(*current_timer) != scope_id) {
-      continue;
+    if (ThreadMatches(thread_id, current_timer) &&
+        capture_data_->ProvideScopeId(*current_timer) == scope_id) {
+      UpdateNextTimerAndGoalTime(&next_timer, goal_time, current_timer, current_time);
     }
-    if ((current_timer->end() > current_time) && (goal_time > current_timer->end())) {
-      next_timer = current_timer;
-      goal_time = current_timer->end();
+  }
+  return next_timer;
+}
+
+const TimerInfo* TimeGraph::FindNextThreadTrackTimer(uint64_t scope_id, uint64_t current_time,
```

Comment:
This should be outside of the TimeGraph, and instead in ThreadTrackDataProvider. But I understand that you are copy pasting the previous solution, so feel free to not address it.